File: openc3/templates/microservice/microservices/TEMPLATE/microservice.rb
# encoding: ascii-8bit

# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/microservices/microservice'
require 'openc3/api/api'

module OpenC3
  class <%= microservice_class %> < Microservice
    include Api # Provides access to api methods

    def initialize(name)
      super(name)
      @config['options'].each do |option|
        # Update with your own OPTION handling
        case option[0].upcase
        when 'PERIOD'
          @period = option[1].to_i
        else
          @logger.error("Unknown option passed to microservice #{@name}: #{option}")
        end
      end

      @period = 60 unless @period # 1 minutes
      @sleeper = Sleeper.new
    end

    def run
      while true
        start_time = Time.now
        break if @cancel_thread

        # Do your microservice work here
        Logger.info("Template Microservice ran")
        # cmd("INST ABORT")

        # The @state variable is set to 'RUNNING' by the microservice base class
        # The @state is reflected to the user in the MICROSERVICES tab so you can
        # convey long running actions by changing it, e.g. @state = 'CALCULATING ...'

        run_time = Time.now - start_time
        delta = @period - run_time
        if delta > 0
          # Delay till the next period
          break if @sleeper.sleep(delta) # returns true and breaks loop on shutdown
        end
        @count += 1
      end
    end

    def shutdown
      @sleeper.cancel # Breaks out of run()
      super()
    end
  end
end

OpenC3::<%= microservice_class %>.run if __FILE__ == $0

File: openc3/templates/microservice/microservices/TEMPLATE/microservice.py
import time
from openc3.microservices.microservice import Microservice
from openc3.utilities.sleeper import Sleeper
from openc3.api import *


class <%= microservice_class %>(Microservice):
    def __init__(self, name):
        super().__init__(name)
        for option in self.config['options']:
            # Update with your own OPTION handling
            match option[0].upper():
                case 'PERIOD':
                    self.period = int(option[1])
                case _:
                    self.logger.error(
                        "Unknown option passed to microservice #{@name}: #{option}"
                    )

        if not self.period:
            self.period = 60  # 1 minutes
        self.sleeper = Sleeper()

    def run(self):
        while True:
            start_time = time.time()
            if self.cancel_thread:
                break

            # Do your microservice work here
            self.logger.info("Template Microservice ran")
            # cmd("INST ABORT")

            # The self.state variable is set to 'RUNNING' by the microservice base class
            # The self.state is reflected to the user in the MICROSERVICES tab so you can
            # convey long running actions by changing it, e.g. self.state = 'CALCULATING ...'

            run_time = time.time() - start_time
            delta = self.period - run_time
            if delta > 0:
                # Delay till the next period
                if self.sleeper.sleep(
                    delta
                ):  # returns true and breaks loop on shutdown
                    break
            self.count += 1

    def shutdown(self):
        self.sleeper.cancel()  # Breaks out of run()
        super().shutdown()


if __name__ == "__main__":
    <%= microservice_class %>.run()
